Provide overloading for Eigen's AutoDiffXd

soonho-tri opened the pull request:
which evaluates terms immediately and returns an AutoDiffXd instead of expression trees.

1. Provide a template specialization, `AutoDiffScalar<VectorXd>`, which modifies the return type of operators (`+`, `-`, `*`, `/`). 
2. Provide overloading of math functions (such as `sin`, `cos`, `pow`, `log`, `exp`, ...) for `AutoDiffXd`.

A consequence of having our own version of `AutoDiffScalar<VectorXd>` is that we need to provide its definition before any code instantiates it. To make sure that, I've replaced `#include <unsupported/Eigen/AutoDiff>` with `#include "drake/common/autodiff_overloads.h"` in this PR.

Fixes #6944

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7007)
<!-- Reviewable:end -->

soonho-tri commented:
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-memcheck-lsan please
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-everything please
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-debug please

soonho-tri commented:
+@jwnimmer-tri , for feature-review please.

---

Review status: 0 of 28 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-:-KtsJJmHkpfvZ8bLbVLb:bwjlzch)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 26 of 28 files at r1.
Review status: 26 of 28 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I am going to stop here for the night, in case you want to shuffle the files around first.

---

Review status: 26 of 28 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/autodiff_overloads.h, line 123 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-KtsmVEg5NqaoEvQJvrK:-KtsmVEg5NqaoEvQJvrL:b-5330sw) ([raw file](https://github.com/robotlocomotion/drake/blob/e5514b658913ea3513cbbac131c0bb0164b0f838/drake/common/autodiff_overloads.h#L123)):*
> ```objc
>           x_to_the_y * log(x) * ygrad);
> }
> ```

The pre-existing content of this file is more or less Drake additions on top of Eigen, whereas this PR is more or less fixing existing Eigen.  I wonder if keeping those two ideas cut into separate files would make sense?

---

*[drake/common/autodiff_overloads.h, line 141 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-KtsjGqVMNF0Gy4zfW7J:-KtsjGqWO8da_l1IJa5-:bpmv702) ([raw file](https://github.com/robotlocomotion/drake/blob/e5514b658913ea3513cbbac131c0bb0164b0f838/drake/common/autodiff_overloads.h#L141)):*
> ```objc
> // The following class definition is a modification from the
> // `unsupported/Eigen/src/AutoDiff/AutoDiffScalar.h` file in Eigen's source
> // code.
> ```

If I recall the MPL correctly, per the MPL we should segregate this class specialization (which is mostly a copy of Eigen code) into a new file with appropriate licensing (covered under MPL not Drake BSD) , and `#include` it here.  Did you look into the rules around this yet?

If you end up doing that, perhaps it would be nifty to start with the first commit of the file (outside of the build system) as unmodified upstream, so that our local diffs are more apparent?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-:-Ktsmj_UhEWJDXOCjAYe:bm27yii)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 26 of 31 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/autodiff_overloads.h, line 123 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-KtsmVEg5NqaoEvQJvrK:-KtwmbGGrwv-bQplgJpd:b-eclw99) ([raw file](https://github.com/robotlocomotion/drake/blob/e5514b658913ea3513cbbac131c0bb0164b0f838/drake/common/autodiff_overloads.h#L123)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The pre-existing content of this file is more or less Drake additions on top of Eigen, whereas this PR is more or less fixing existing Eigen.  I wonder if keeping those two ideas cut into separate files would make sense?
</blockquote></details>

Done. FYI, this file is not changed anymore, except the `#include`s.

---

*[drake/common/autodiff_overloads.h, line 141 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-KtsjGqVMNF0Gy4zfW7J:-Ktwm_kHXxhZSaXmIfIZ:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/e5514b658913ea3513cbbac131c0bb0164b0f838/drake/common/autodiff_overloads.h#L141)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

If I recall the MPL correctly, per the MPL we should segregate this class specialization (which is mostly a copy of Eigen code) into a new file with appropriate licensing (covered under MPL not Drake BSD) , and `#include` it here.  Did you look into the rules around this yet?

If you end up doing that, perhaps it would be nifty to start with the first commit of the file (outside of the build system) as unmodified upstream, so that our local diffs are more apparent?
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
@drake-jenkins-bot linux-xenial-gcc-ninja-experimental-open-source-debug please

---

Review status: 26 of 32 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-:-Kty7-_7Cu_-HuG0QRBT:b-ioqduq)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 6 of 6 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-Le_p7-2j0XWWIwjZ:-Ku-Le_qUJrwkps56m20:btdpku9) (no related file):*
FYI It would be ideal if none of the commits beyond the first showed a file coming in as-new, but rather that every commit was diffs of some kind, so that `git blame` is more precise.  I believe you could curate the commits even more in order to convince git to recognize the diffs.  

Instead of both moving the file and editing it in the same (currently second) commit, I would suggest either that the first commit place the file with its final name, or to insert a rename-only commit before editing begins (between the currently first and second commits).

Also, it might make sense to commit the reformatting change (de-indent by 2, clang-format linewraps) separately from the functional change (of changing return types from references or Cwise, to AutoDiffXd).  Backing out those two stages (locally in emacs) is essentially what I had to do to review the code anyway.  This is both more likely to convince git to show things reasonably, and provide better clarity to future readers on exactly what we (meaningfully) changed.

In short, my curation suggestion is something like:
- add upstream file (same as now)
- rename file into final home
- non-function reformatting, indentation, etc.
- strip out unwanted bits, alter wanted bits, add unit test, add new component into build
- port Drake to use it (same as now)

---

*[drake/common/autodiffxd.h, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-GG7N6hps4tSQNepu:-Ku-GG7N6hps4tSQNepv:bmio542) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L22)):*
> ```objc
> // Explicit template specializations of Eigen::AutoDiffScalar for VectorXd.
> //
> // AutoDiffScalar (ADS) tries to call internal::make_coherent to promote empty
> ```

FYI We don't appear to use the "ADS" abbreviation anywhere later on in the docs.

---

*[drake/common/autodiffxd.h, line 69 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-M46S6GRAURVBqrwc:-Ku-M46S6GRAURVBqrwd:bkqkaig) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L69)):*
> ```objc
>               Scalar, typename internal::traits<typename internal::remove_all<
>                           OtherDerType>::type>::Scalar>::value &&
>               internal::is_convertible<OtherDerType, DerType>::value,
> ```

Could you comment why you needed to add this second clause to the conjunction here, whereas upstream did not?

---

*[drake/common/autodiffxd.h, line 298 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-OZey-nJFwnE6VeRZ:-Ku-OZey-nJFwnE6VeR_:b-jolv7q) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L298)):*
> ```objc
> 
> DRAKE_EIGEN_AUTODIFFXD_DECLARE_GLOBAL_BINARY(
>     pow, internal::make_coherent(a.derivatives(), b.derivatives());
> ```

FYI Consider moving the `make_coherent` inside the macro, instead of here?  It seems like all cases would need it.

---

*[drake/common/autodiffxd.h, line 298 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-OsrJes3sj2NIzR7h:-Ku-OsrJes3sj2NIzR7i:b-eri5u7) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L298)):*
> ```objc
> 
> DRAKE_EIGEN_AUTODIFFXD_DECLARE_GLOBAL_BINARY(
>     pow, internal::make_coherent(a.derivatives(), b.derivatives());
> ```

Do we need to declare this?  It looks to me like Drake's `autodiff_overloads.h` for `pow(ADS, ADS)` should already be correct -- it returns by-value.

---

*[drake/common/autodiffxd.h, line 322 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-PJ9fRlRrPjsHYlRS:-Ku-PJ9fRlRrPjsHYlRT:b-3gvxb5) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L322)):*
> ```objc
>                          a.value() * b.derivatives()) /
>                         squared_hypot;
>     return ret;)
> ```

Do we need to declare this?  It looks to me like `AutoDiffScalar.h` already returns by-value in the case of `atan2(ADS, ADS)`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 29 of 32 files reviewed at latest revision, 6 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-Le_p7-2j0XWWIwjZ:-Ku0FVQD0QV0PAGYFxtM:b4o4mw1) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI It would be ideal if none of the commits beyond the first showed a file coming in as-new, but rather that every commit was diffs of some kind, so that `git blame` is more precise.  I believe you could curate the commits even more in order to convince git to recognize the diffs.  

Instead of both moving the file and editing it in the same (currently second) commit, I would suggest either that the first commit place the file with its final name, or to insert a rename-only commit before editing begins (between the currently first and second commits).

Also, it might make sense to commit the reformatting change (de-indent by 2, clang-format linewraps) separately from the functional change (of changing return types from references or Cwise, to AutoDiffXd).  Backing out those two stages (locally in emacs) is essentially what I had to do to review the code anyway.  This is both more likely to convince git to show things reasonably, and provide better clarity to future readers on exactly what we (meaningfully) changed.

In short, my curation suggestion is something like:
- add upstream file (same as now)
- rename file into final home
- non-function reformatting, indentation, etc.
- strip out unwanted bits, alter wanted bits, add unit test, add new component into build
- port Drake to use it (same as now)
</blockquote></details>

Thanks for the suggestions. I followed them. PTAL.

---

*[drake/common/autodiffxd.h, line 22 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-GG7N6hps4tSQNepu:-Ku-tArUBgfV1ZM5BBcY:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L22)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI We don't appear to use the "ADS" abbreviation anywhere later on in the docs.
</blockquote></details>

Done.

---

*[drake/common/autodiffxd.h, line 69 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-M46S6GRAURVBqrwc:-Ku-tUJjVl_DGKNo39_c:b-3xd8tb) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L69)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Could you comment why you needed to add this second clause to the conjunction here, whereas upstream did not?
</blockquote></details>

Removed.

You're correct. `Eigen-3.3.3`, which is the one that we're using, does not have the conjunction. 

FYI, the conjunction is added by https://bitbucket.org/eigen/eigen/commits/c7cf575724de40626aedb175ba0d6bd39bbfcac2 and is used in `>=Eigen-3.3.4`. Since we're using `3.3.3`, I think it's fair to keep up with that version.

---

*[drake/common/autodiffxd.h, line 298 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-OZey-nJFwnE6VeRZ:-Ku09sE2c5p0QzFJNmcO:b8e34us) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L298)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI Consider moving the `make_coherent` inside the macro, instead of here?  It seems like all cases would need it.
</blockquote></details>

Done. (not using macro anymore)

---

*[drake/common/autodiffxd.h, line 298 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-OsrJes3sj2NIzR7h:-Ku09w-ICGIUvxpODekv:bw4wujh) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L298)):*
> Do we need to declare this? 

You're right. We don't need this anymore.

> Drake's autodiff_overloads.h for pow(ADS, ADS) should already be correct

No, it's not correct. There were two problems:

 - It didn't call `make_coherent`
 - Its return type and `static_check` is not precise and didn't work for some corner cases. For example, `pow(x + 1.0, y)` didn't work when `x` and `y` are `AutoDiff3d`. Because `x + 1.0` returns `ADS(DerType&)` not `ADS(DerType)` and we can't access`DerType&::PlainObject` (while we can for `DerType::PlainObject`).

I've fixed those problems in `autodiff_overloads.h`.

---

*[drake/common/autodiffxd.h, line 322 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-PJ9fRlRrPjsHYlRS:-Ku0Fiz31tOMwil9mVjE:b-3feo4e) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L322)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Do we need to declare this?  It looks to me like `AutoDiffScalar.h` already returns by-value in the case of `atan2(ADS, ADS)`.
</blockquote></details>

I've found that we still need this just because Eigen's version does not call `make_coherent`. Since `atan2` is the only binary math-function that we're adding for `(ADS, ADS)` case, I've added its implementation at the end of the file without using a macro.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 29 of 32 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/autodiffxd.h, line 322 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-PJ9fRlRrPjsHYlRS:-Ku0VJ10w10N4RiBjewX:b-fq8e41) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L322)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I've found that we still need this just because Eigen's version does not call `make_coherent`. Since `atan2` is the only binary math-function that we're adding for `(ADS, ADS)` case, I've added its implementation at the end of the file without using a macro.
</blockquote></details>

Just in case, here is the link to the Eigen-3.3.3's `atan2` implementation for ADS:
https://bitbucket.org/eigen/eigen/src/3.3.3/unsupported/Eigen/src/AutoDiff/AutoDiffScalar.h?at=default#AutoDiffScalar.h-620

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
LGTM at head, but I want to check on the curation before posting the emoji.  Ready for platform review, though.

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/autodiffxd.h, line 322 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-PJ9fRlRrPjsHYlRS:-Ku0Igkjovy6DGo7vLDu:bbn2h8b) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L322)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Just in case, here is the link to the Eigen-3.3.3's `atan2` implementation for ADS:
https://bitbucket.org/eigen/eigen/src/3.3.3/unsupported/Eigen/src/AutoDiff/AutoDiffScalar.h?at=default#AutoDiffScalar.h-620
</blockquote></details>

Ah.  Maybe add a comment at our `atan2` below then, clarifying why we're doing it for future readers?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-:-Ku0eiLQVnrSfJdIKqGl:b5katl8)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
+@ericcousineau-tri , for platform review. Thank you!

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-:-Ku0xRZi9zonyDbjNoFh:b-jc3f4l)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 31 of 32 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/autodiffxd.h, line 322 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku-PJ9fRlRrPjsHYlRS:-Ku1-607Ck12xsn7VxUE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/c181e435175308d102beff1397b91945d9b53492/drake/common/autodiffxd.h#L322)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Ah.  Maybe add a comment at our `atan2` below then, clarifying why we're doing it for future readers?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-:-Ku106ZIIKJhXubXd_o8:bnfp4nl)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 26 of 28 files at r1, 3 of 6 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/bindings/pybind11/pydrake_autodiff_types.h, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0xHlanbuu-B_hVvko:-Ku0xHlbWh4yU2sTxkWT:b-8y4jvs) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/bindings/pybind11/pydrake_autodiff_types.h#L6)):*
> ```objc
> #include "drake/common/autodiff_overloads.h"
> 
> typedef Eigen::AutoDiffScalar<Eigen::VectorXd> AutoDiffXd;
> ```

BTW This is unrelated, but why do we redefine this typedef from `drake/eigen_autodiff_types.h`?
Should we consolidate those types, possibly moving the types below into that header?

---

*[drake/common/autodiffxd.h, line 3 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0yW8zzYn7AqJKyVjF:-Ku0yW9-7tP6QH3FAz12:bpqoo93) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/autodiffxd.h#L3)):*
> ```objc
> #pragma once
> 
> // This file is a modification of Eigen's AutoDiffScalar.h file.
> ```

Could you add in that this taken from Eigen v3.3.3?

---

*[drake/common/test/autodiffxd_test.cc, line 16 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku11HPbKGLaHyIXVFU_:-Ku11HPchoXNzq6f1J2S:b-sbv8m2) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L16)):*
> ```cpp
>     // We only set the derivatives of y and leave x's uninitialized.
>     y_xd_.derivatives() = Eigen::VectorXd::Ones(3);
>     y_3d_.derivatives() = Eigen::VectorXd::Ones(3);
> ```

BTW nit Any reason that you use `VectorXd` rather than `Vector3d` for `y_3d_`?

---

*[drake/common/test/autodiffxd_test.cc, line 570 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku12uNyEvIyXWSm9oDQ:-Ku12uNyEvIyXWSm9oDR:b8l6863) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L570)):*
> ```cpp
>       }));  // 0.5 -x + y
>   EXPECT_TRUE(
>       Check([](const auto& x, const auto& y) -> DRAKE_NO_REF_ADS_TYPE(x) {
> ```

It looks like all of these checks could be reduced a more generally scoped macro:
```
#define CHECK_STUFF(expr) \
  EXPECT_TRUE( \
    Check([](const auto& x, const auto& y) -> DRAKE_NO_REFS_ADS_TYPE(x) { \
       return expr; \
    })) << #expr  // Print statement to locate it if it fails
...

CHECK_STUFF(cosh(x * x) + y);
CHECK_STUFF(max(0.6 - y, 0.6 - x) + y);
```
Could it be handled by a quick regex replace?

Also, it seems like some of these patterns are slightly arbitrary, and spread out across the different operators / functions.
Could you consolidated it to something like {unary, binary} + {operator, function} and use the exact same variants, to ensure that everything is expansively tested?

`CHECK_BINARY_FUNCTION(f)`  --> will check f(y, x), f(y + y, x + x), f(3 * -y, 1 / x) + x, etc.

---

*[drake/common/test/autodiffxd_test.cc, line 573 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku4tJezjyYpj8CbflxR:-Ku4tJf-_8GSrhhEdXNz:b10xveq) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L573)):*
> ```cpp
>         return max(0.6 - y, 0.6 - x) + y;
>       }));  // 0.6 -x + y
> }
> ```

Is there any way that you could add a negative test?
e.g., somehow before the specialization is included, show that `CHECK` fails?
Or make a utility file with the checking code (let it violate IWYU, if necessary), and then have two tests files, where (a) one does not include the specialization and (b) one does?

---

*[drake/common/test/autodiffxd_test.cc, line 575 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku4tLLHImpOEPSGy7jL:-Ku4tLLHImpOEPSGy7jM:b-7zwdd) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L575)):*
> ```cpp
> }
> 
> #undef DRAKE_NO_REF_ADS_TYPE
> ```

BTW Is this unnecessary? Its scope dies at the end of this translation unit.

---

*[drake/math/BUILD, line 154 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku14uD8Qhg7Y6PRPODb:-Ku14uD8Qhg7Y6PRPODc:bad0n8c) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/math/BUILD#L154)):*
> ```Python
>     deps = [
>         "//drake/common",
>         "//drake/common:autodiff",
> ```

BTW Rather than have `:autodiff` be separate from `common`, could we include this into `common` overall?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, pending one comment on generalizing / making the tests more consistent, and possibly adding negative tests.

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-:-Ku4tWnKmqBm3CIY_aFY:bvqdnl9)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/common/autodiffxd.h, line 3 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0yW8zzYn7AqJKyVjF:-Ku5-Nbr6pXFHfWaPQ1I:b-sbpfjf) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/autodiffxd.h#L3)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Could you add in that this taken from Eigen v3.3.3?
</blockquote></details>

FYI Soonho maybe just amend the first commit message?  The full commit sha there is awesome and should remain, but doesn't highlight that its the 3.3.3 release.

---

*[drake/common/test/autodiffxd_test.cc, line 575 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku4tLLHImpOEPSGy7jL:-Ku5-sh08LuSxpgh19o_:bvbyht4) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L575)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Is this unnecessary? Its scope dies at the end of this translation unit.
</blockquote></details>

GSG says _`#define` macros right before you use them, and #undef them right after_. -- https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros, so this is correct.

---

*[drake/math/BUILD, line 154 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku14uD8Qhg7Y6PRPODb:-Ku503Jpv7Jf4ZBg5RMk:b-tk164k) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/math/BUILD#L154)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW Rather than have `:autodiff` be separate from `common`, could we include this into `common` overall?
</blockquote></details>

No, the library currently named `//drake/common:common` must remain as minimal as possible.  See recent `#bazel` slack for a renaming proposal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Fixed easier ones. There will be another commit improving tests.

---

Review status: 30 of 32 files reviewed at latest revision, 6 unresolved discussions.

---

*[drake/bindings/pybind11/pydrake_autodiff_types.h, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0xHlanbuu-B_hVvko:-Ku4yKHYIH5t0MiM-BEC:b8ujrh0) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/bindings/pybind11/pydrake_autodiff_types.h#L6)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW This is unrelated, but why do we redefine this typedef from `drake/eigen_autodiff_types.h`?
Should we consolidate those types, possibly moving the types below into that header?
</blockquote></details>

I agree. 

Just in case, I'm paging @rdeits, the author of this portion of this file.

---

*[drake/common/autodiffxd.h, line 3 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0yW8zzYn7AqJKyVjF:-Ku4z9CJ2XvVT9bQK-S6:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/autodiffxd.h#L3)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Could you add in that this taken from Eigen v3.3.3?
</blockquote></details>

Done.

---

*[drake/common/test/autodiffxd_test.cc, line 16 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku11HPbKGLaHyIXVFU_:-Ku4zB8h9ru9ziuabtp1:bb70bx7) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L16)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW nit Any reason that you use `VectorXd` rather than `Vector3d` for `y_3d_`?
</blockquote></details>

Done. (Changed to `Vector3d`).

---

*[drake/common/test/autodiffxd_test.cc, line 575 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku4tLLHImpOEPSGy7jL:-Ku4zgAbliHgfD4NRv-m:b-df126u) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L575)):*
[Google CPP guide](https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros) says the following:

> - Don't define macros in a .h file.
> - #define macros right before you use them, and #undef them right after.

and I think we need it here to be consistent with it. In the worst case, there is nothing wrong with it :-)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-:-Ku5123Hxx-i-DgMSmw_:bldctpm)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 30 of 32 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/test/autodiffxd_test.cc, line 575 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku4tLLHImpOEPSGy7jL:-Ku52H8YYCxLvQ72WLCi:bht0gu9) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L575)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

[Google CPP guide](https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros) says the following:

> - Don't define macros in a .h file.
> - #define macros right before you use them, and #undef them right after.

and I think we need it here to be consistent with it. In the worst case, there is nothing wrong with it :-)

</blockquote></details>

Gotcha, thank you!

---

*[drake/math/BUILD, line 154 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku14uD8Qhg7Y6PRPODb:-Ku52N4yJwyoLPr80XfC:b48k9my) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/math/BUILD#L154)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

No, the library currently named `//drake/common:common` must remain as minimal as possible.  See recent `#bazel` slack for a renaming proposal.
</blockquote></details>

Gotcha, I've added in my thumbs-up on the discussion as well. Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 30 of 32 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/common/test/autodiffxd_test.cc, line 573 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku4tJezjyYpj8CbflxR:-Ku51Qn_DSuMKTHAJ8Yg:b-gf4iab) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L573)):*
> where (a) one does not include the specialization and (b) one does?

Hmm.. FYI, I'll work on https://github.com/RobotLocomotion/drake/issues/7015, which will introduce the representative header for AutoDiff. That will make it quite hard to follow your proposal.

We're fixing problems in Eigen and I don't think it has high value to show the existence of the problems in the tests. How about having a `TODO` saying "When we upgrade our Eigen, please check if we still need this specialization"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/bindings/pybind11/pydrake_autodiff_types.h, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0xHlanbuu-B_hVvko:-Ku52nQsx7u4xZ9sUUbk:b-ywv6b2) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/bindings/pybind11/pydrake_autodiff_types.h#L6)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

I agree. 

Just in case, I'm paging @rdeits, the author of this portion of this file.
</blockquote></details>

FYI I think reusing `drake::AutoDiffXd` makes sense, unless there is some headache.

Personally, I would prefer _not_ to move the other typedefs from here into common unless there are non-pybind11 files that need them.

---

*[drake/common/autodiffxd.h, line 3 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0yW8zzYn7AqJKyVjF:-Ku537qhouFO6InELNIK:b-rqqg52) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/autodiffxd.h#L3)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think perhaps your clipboard pasted the wrong URL.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 31 of 32 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/common/autodiffxd.h, line 3 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0yW8zzYn7AqJKyVjF:-Ku53g13HTe7IHxY46zH:bpckh81) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/autodiffxd.h#L3)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I think perhaps your clipboard pasted the wrong URL.
</blockquote></details>

Done. (Thanks!!)

---

*[drake/common/test/autodiffxd_test.cc, line 573 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku4tJezjyYpj8CbflxR:-Ku54XFjxMXUSA4IMFln:blv2t8p) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L573)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

> where (a) one does not include the specialization and (b) one does?

Hmm.. FYI, I'll work on https://github.com/RobotLocomotion/drake/issues/7015, which will introduce the representative header for AutoDiff. That will make it quite hard to follow your proposal.

We're fixing problems in Eigen and I don't think it has high value to show the existence of the problems in the tests. How about having a `TODO` saying "When we upgrade our Eigen, please check if we still need this specialization"?
</blockquote></details>

FYI, a `TODO` is added to `autodiffxd.h` file.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

rdeits commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/bindings/pybind11/pydrake_autodiff_types.h, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0xHlanbuu-B_hVvko:-Ku5HuuyI8w2a2djYCmV:bs1qrl) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/bindings/pybind11/pydrake_autodiff_types.h#L6)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I think reusing `drake::AutoDiffXd` makes sense, unless there is some headache.

Personally, I would prefer _not_ to move the other typedefs from here into common unless there are non-pybind11 files that need them.
</blockquote></details>

Probably just because I didn't know that definition existed. Should be safe to remove this duplicate.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/bindings/pybind11/pydrake_autodiff_types.h, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0xHlanbuu-B_hVvko:-Ku5IMCWuP56NdgCWukt:b-ojhs1r) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/bindings/pybind11/pydrake_autodiff_types.h#L6)):*
<details><summary><i>Previously, rdeits (Robin Deits) wrote…</i></summary><blockquote>

Probably just because I didn't know that definition existed. Should be safe to remove this duplicate.
</blockquote></details>

@rdeits , thanks for the confirmation. I'll take care of it in this PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

rdeits commented:
By the way, the `bindings/pybind11/pydrake_autodiffutils.cc` file exists entirely to force Eigen to evaluate autodiff expressions immediately before returning them to python (since Python doesn't know what to do with an expression tree). If this PR is making that behavior the default for AutoDiffXd, then it might be possible to substantially simplify `pydrake_autodiffutils.cc`. See https://github.com/RobotLocomotion/drake/blob/master/drake/bindings/pybind11/pydrake_autodiffutils.cc#L20 and all of the lambdas in the rest of that file which just call `eval()` on their results.

soonho-tri commented:
@rdeits , thanks for the information! After this PR gets merged, I'll check and open another PR if it turns out we don't need those code in `pydrake_autodiffutils.cc`.

soonho-tri commented:
Review status: 27 of 34 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/bindings/pybind11/pydrake_autodiff_types.h, line 6 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku0xHlanbuu-B_hVvko:-Ku6oBcrRX1YSawHJEG2:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/bindings/pybind11/pydrake_autodiff_types.h#L6)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

@rdeits , thanks for the confirmation. I'll take care of it in this PR.
</blockquote></details>

Done.

---

*[drake/common/test/autodiffxd_test.cc, line 570 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku12uNyEvIyXWSm9oDQ:-Ku6oFhJmt9SfJ9UwKFm:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L570)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

It looks like all of these checks could be reduced a more generally scoped macro:
```
#define CHECK_STUFF(expr) \
  EXPECT_TRUE( \
    Check([](const auto& x, const auto& y) -> DRAKE_NO_REFS_ADS_TYPE(x) { \
       return expr; \
    })) << #expr  // Print statement to locate it if it fails
...

CHECK_STUFF(cosh(x * x) + y);
CHECK_STUFF(max(0.6 - y, 0.6 - x) + y);
```
Could it be handled by a quick regex replace?

Also, it seems like some of these patterns are slightly arbitrary, and spread out across the different operators / functions.
Could you consolidated it to something like {unary, binary} + {operator, function} and use the exact same variants, to ensure that everything is expansively tested?

`CHECK_BINARY_FUNCTION(f)`  --> will check f(y, x), f(y + y, x + x), f(3 * -y, 1 / x) + x, etc.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 7 of 7 files at r7.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/test/autodiffxd_test.cc, line 294 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku6x3s-1c4sZKOxn94n:-Ku6x3s-1c4sZKOxn94o:b-uk7qeu) ([raw file](https://github.com/robotlocomotion/drake/blob/47f19a59850c82dfa2f509b99b00bc72892a2b0b/drake/common/test/autodiffxd_test.cc#L294)):*
> ```cpp
> 
> #undef CHECK_BINARY_FUNCTION_ADS_ADS
> #undef CHECK_BINARY_FUNCTION_ADS_SCALAR
> ```

FYI I think you can swap Pow1 case up before the ADS_SCALAR macro, and then undef ADS_ADS earlier on?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: 33 of 34 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/common/test/autodiffxd_test.cc, line 294 at r7](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku6x3s-1c4sZKOxn94n:-Ku71o93ZxN7mHoYRAH1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/47f19a59850c82dfa2f509b99b00bc72892a2b0b/drake/common/test/autodiffxd_test.cc#L294)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI I think you can swap Pow1 case up before the ADS_SCALAR macro, and then undef ADS_ADS earlier on?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

soonho-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/test/autodiffxd_test.cc, line 570 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku12uNyEvIyXWSm9oDQ:-KuJxbsXaOw0HVO2lgZO:b-4ey4jg) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L570)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

@ericcousineau-tri , could you check this one when you have time? Thanks for the suggestion by the way, it looks cleaner now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Reviewed 6 of 7 files at r7, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/common/test/autodiffxd_test.cc, line 570 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/7007#-Ku12uNyEvIyXWSm9oDQ:-KuKGyplOVw7cE-u75A3:bwga3te) ([raw file](https://github.com/robotlocomotion/drake/blob/b61ad1c433fba6ac18e60edf28d43e0753c868c6/drake/common/test/autodiffxd_test.cc#L570)):*
<details><summary><i>Previously, soonho-tri (Soonho Kong) wrote…</i></summary><blockquote>

@ericcousineau-tri , could you check this one when you have time? Thanks for the suggestion by the way, it looks cleaner now.
</blockquote></details>

Looks awesome, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7007)*
<!-- Sent from Reviewable.io -->
